pkg/operator/encryption/statemachine/transition.go

```diff
@@ -0,0 +1,238 @@
+package statemachine
+
+import (
+	"fmt"
+
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	apiserverconfigv1 "k8s.io/apiserver/pkg/apis/config/v1"
+	corev1client "k8s.io/client-go/kubernetes/typed/core/v1"
+	"k8s.io/klog"
+
+	"github.com/openshift/cluster-kube-apiserver-operator/pkg/operator/encryption/encryptionconfig"
+	"github.com/openshift/cluster-kube-apiserver-operator/pkg/operator/encryption/secrets"
+	"github.com/openshift/cluster-kube-apiserver-operator/pkg/operator/encryption/state"
+	"github.com/openshift/cluster-kube-apiserver-operator/pkg/operator/operatorclient"
+)
+
+func GetEncryptionConfigAndState(
+	podClient corev1client.PodsGetter,
+	secretClient corev1client.SecretsGetter,
+	targetNamespace string,
+	encryptionSecretSelector metav1.ListOptions,
+	encryptedGRs []schema.GroupResource,
+) (current *apiserverconfigv1.EncryptionConfiguration, desired map[schema.GroupResource]state.GroupResourceState, secretsFound bool, transitioningReason string, err error) {
+	revision, err := getAPIServerRevisionOfAllInstances(podClient.Pods(targetNamespace))
+	if err != nil {
+		return nil, nil, false, "", err
+	}
+	if len(revision) == 0 {
+		return nil, nil, false, "APIServerRevisionNotConverged", nil
+	}
+
+	encryptionConfig, err := getCurrentEncryptionConfig(secretClient.Secrets(targetNamespace), revision)
+	if err != nil {
+		return nil, nil, false, "", err
+	}
+
+	encryptionSecretList, err := secretClient.Secrets(operatorclient.GlobalMachineSpecifiedConfigNamespace).List(encryptionSecretSelector)
+	if err != nil {
+		return nil, nil, false, "", err
+	}
+	var encryptionSecrets []*corev1.Secret
+	for i, s := range encryptionSecretList.Items {
+		if _, err := secrets.ToKeyState(&s); err != nil {
```

Comment:
yeah, it reads well, it tells me immediately that the secret will be transformed into a keyState